### PR TITLE
Adding upcoming gauges query

### PR DESCRIPTION
**Pull request by xBalbinus**
Closes: #1158 

## Description

Added new request definitions in the proto. UpcomingGaugesPerDenom, UpcomingGaugesPerDenomRequest, UpcomingGaugesPerDenomResponse

Made proto-all

Added a keeper method in x/incentives/keeper/grpc_query.go .

Ditto for adding a test for this in grpc_query_test.go: note: I implemented tests as well for querying active upcoming gauges per denom. At the moment, they're not table driven but still diving deeper in and will implement soon.

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/gaia/blob/main/CONTRIBUTING.md#pr-targeting))
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: proto/osmosis/incentives/query.proto
Comment (ValarDragon) [inline proto/osmosis/incentives/query.proto:140]:
```suggestion
  repeated Gauge upcoming_gauges = 1 [ (gogoproto.nullable) = false ];
```

**Comment by ValarDragon** [comment]
Nice work! I think the `mytestnet` folder got accidentally committed

File: x/incentives/keeper/grpc_query.go
Comment (ValarDragon) [inline x/incentives/keeper/grpc_query.go:153]: On reviewing this PR, I'm realizing that every filtered paginate usage in this package is wrong. I think its fine to merge this as is though, since its not a real performance overhead / can be fixed later. (We should make an issue to track it)

Filtered paginate's method of working is actually pretty awkward / maybe we need to change it at the SDK level. It gives you the value in `[]bytes`, that this code should then de-marshal, and check if it meets the required constraints. If accumulate is true, only then add it to the gauges result. But this is a high amount of overhead tbh. I think its probably worth getting this refactored in the SDK to be simple, with go 1.18 Generics.

I'll add more detail on this / convert it to an issue, doesn't block this PR!

**Review by ValarDragon — APPROVED** [review]
Nice job on your first code change PR!!

LGTM! (It has the same behavior as ActiveGaugesPerDenom)

We should make two follow on issues:
* One brainstorming ways for fixing the erroneous pagination behavior (either we change things directly here, or add  better pagination code scaffolding)
* Steps for getting a better table driven test framework here. (We need to make it easy to make upcoming, present, and completed gauges in the test suite)

File: x/incentives/keeper/grpc_query_test.go
Comment (ValarDragon) [inline x/incentives/keeper/grpc_query_test.go:115]: oh you have to make a querier struct now for this to work now. (An update happened on main. 

Change to `suite.querier,ActiveGaugesPerDenom` and I think that should work in all places where this is erroring.

**Comment by ValarDragon** [comment]
Can you run `make format` to fix the lint errors

**Comment by xBalbinus** [comment]
Sweet, got you, checking it out!

**Comment by ValarDragon** [comment]
PR LGTM, its having the liveness test timeout, but this confuses me, since this doesn't even touch the state machine? Any ideas @alexanderbez ?

File: proto/osmosis/incentives/query.proto
Comment (alexanderbez) [inline proto/osmosis/incentives/query.proto:53]: nit: add a godoc for this handler 👍

Comment (alexanderbez) [reply proto/osmosis/incentives/query.proto:53]: Also, why is it structured like this? Makes it very hard to read. Did we not add the vscode proto formatting changes @ValarDragon?

File: x/incentives/client/cli/query.go
Comment (alexanderbez) [inline x/incentives/client/cli/query.go:326]: nit: for concise/straight forward CLI handlers like this, I'd argue `Long` is not necessary as it does not offer any additional help or insight.

File: x/incentives/keeper/grpc_query.go
Comment (alexanderbez) [inline x/incentives/keeper/grpc_query.go:150]:
```suggestion
	prefixStore := prefix.NewStore(store, types.KeyPrefixUpcomingGauges)
```

File: x/incentives/keeper/grpc_query.go
Comment (alexanderbez) [inline x/incentives/keeper/grpc_query.go:159]: Hmmm, why are we returning true here?

File: x/incentives/keeper/grpc_query.go
Comment (alexanderbez) [inline x/incentives/keeper/grpc_query.go:147]: Nit: I know the other queries don't do this, which we need to address IMO, but we should have an explicit `nil` check here for `req`.

i.e.
```go
if req == nil {
  	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "empty request")
	}
}

	if req.Denom == "" {
		return nil, status.Error(codes.InvalidArgument, "invalid denom")
	}

```

Comment (alexanderbez) [reply x/incentives/keeper/grpc_query.go:147]: Granted, we probably shouldn't use `InvalidArgument` for `req` being nil, but rather something else.

**Review by alexanderbez — CHANGES_REQUESTED** [review]
Looks great @xBalbinus!

I left some minor feedback. I'm also taking a look at the liveness test.

File: proto/osmosis/incentives/query.proto
Comment (ValarDragon) [reply proto/osmosis/incentives/query.proto:53]: Theres no vscode proto formatting AFAIK. I'm down for us to add it. (IIRC you did it via clang, not sure how that install works across machines / operating systems)

Comment (alexanderbez) [reply proto/osmosis/incentives/query.proto:53]: https://github.com/osmosis-labs/osmosis/pull/1221

:P

**Comment by alexanderbez** [comment]
Re liveness failures:

```
➜ starport chain serve --reset-once -v -c ./starport.ci.yml
[STARPORT] Cosmos SDK's version is: stargate - v0.45.1
[STARPORT] 
[STARPORT] 🛠️  Building proto...
[STARPORT] cannot build app:
[STARPORT] 
[STARPORT]      cannot build app:
[STARPORT] 
[STARPORT]      error while running command /var/folders/jl/532qskgn2y32m_qg3p4l8q440000gn/T/protoc832737936: osmosis/incentives/query.proto:136:54: Field number 1 has already been used in "osmosis.incentives.UpcomingGaugesPerDenomRequest" by field "denom".
[STARPORT] : exit status 1
```


Just an FYI, if liveness fails or timesout, you can always run it locally to find out:

```
$ starport chain serve --reset-once -v -c ./starport.ci.yml
```

File: proto/osmosis/incentives/query.proto
Comment (alexanderbez) [inline proto/osmosis/incentives/query.proto:136]: We gotta be more careful with reviewing proto structs.

```suggestion
  string denom = 1;
  cosmos.base.query.v1beta1.PageRequest pagination = 2;
```

**Review by alexanderbez — CHANGES_REQUESTED** [review]

File: proto/osmosis/incentives/query.proto
Comment (ValarDragon) [reply proto/osmosis/incentives/query.proto:53]: ah cool! Make proto-format should fix this as well.

@xBalbinus can you run `make proto-format` and commit that change before we merge this in

File: proto/osmosis/incentives/query.proto
Comment (xBalbinus) [reply proto/osmosis/incentives/query.proto]: Oop! Updated.

Comment (ValarDragon) [reply proto/osmosis/incentives/query.proto]: Yeah your totally right. Is there a way we can lint for this more easily? I don't get why make proto-all even compiled with this as an issue.

File: x/incentives/keeper/grpc_query.go
Comment (ValarDragon) [reply x/incentives/keeper/grpc_query.go:147]: Hrmm, lets make a second issue for this? It'd be nice if theres something we can do one layer up to save if `if req == nil` in every single querier logic call

Comment (alexanderbez) [reply x/incentives/keeper/grpc_query.go:147]: Sounds good! If you create it, you can assign it to me and I'll knock it out 👍

**Comment by xBalbinus** [comment]
make proto-format done; changes committed! Thanks @alexanderbez and @ValarDragon for your input & help.

File: x/incentives/keeper/grpc_query_test.go
Comment (ValarDragon) [inline x/incentives/keeper/grpc_query_test.go:126]: This is failing right now, which is actually correct behavior at the moment. This is because gauges only move from upcoming -> active during the epoch logic at the moment. (We should change the state machine to make it possible for gauges to start out as active, not always as upcoming)

You should call `k.BeginDistribution(suite,ctx, gauge)` before this line to make it work. (Also define the gauge return value from `suite.SetupNewGauge`

We should do the same thing to fix the UpcomingGaugesPerDenom test

File: x/incentives/keeper/grpc_query.go
Comment (ValarDragon) [reply x/incentives/keeper/grpc_query.go:147]: Issue made #1224

File: x/incentives/client/cli/query.go
Comment (alexanderbez) [reply x/incentives/client/cli/query.go:327]: bump @xBalbinus

File: x/incentives/keeper/grpc_query.go
Comment (alexanderbez) [reply x/incentives/keeper/grpc_query.go:159]: bump @xBalbinus

File: x/incentives/keeper/grpc_query.go
Comment (alexanderbez) [inline x/incentives/keeper/grpc_query.go:147]: Now that we've cleaned up the queries nil checking, we should add it here.

File: x/incentives/keeper/grpc_query.go
Comment (ValarDragon) [inline x/incentives/keeper/grpc_query.go:151]: double if req == nil, lets remove the inner if statement

**Review by alexanderbez — APPROVED** [review]
🚀🚀🚀🚀🚀🚀🚀🚀🚀

**Comment by alexanderbez** [comment]
I'm confused as to how CI is passing. Starport liveness test reports:

```
[STARPORT] # github.com/osmosis-labs/osmosis/v7/x/incentives/keeper
[STARPORT] ../../x/incentives/keeper/grpc_query.go:173:47: unknown field 'Data' in struct literal of type "github.com/osmosis-labs/osmosis/v7/x/incentives/types".UpcomingGaugesPerDenomResponse
[STARPORT] : exit status 2
```

**Comment by xBalbinus** [comment]
Should we fix the liveliness tests first before committing? What might be causing this?

**Comment by alexanderbez** [comment]
> Should we fix the liveliness tests first before committing? What might be causing this?

@xBalbinus the error above tells you:

```
[STARPORT] ../../x/incentives/keeper/grpc_query.go:173:47: unknown field 'Data' in struct literal of type "github.com/osmosis-labs/osmosis/v7/x/incentives/types".UpcomingGaugesPerDenomResponse

```

**Comment by xBalbinus** [comment]
I just pushed out some fixes / changes & make all has been doing well for me (all tests passed). Let me know if the liveliness is still an issue.

**Comment by ValarDragon** [comment]
So to fix that liveness error, can you do `make proto-all`, and then replace the usage of Data on line 173, with `UpcomingGauges` which is what should get written after `make proto-all` is ran? That should hopefully fix the liveness test

**Pull request merged**
